probe-rs/src/architecture/arm/communication_interface.rs

```diff
@@ -328,10 +340,33 @@ pub struct MemoryApInformation {
 /// (E.g. CMSIS-DAP and J-Link support this, ST-Link does not)
 #[derive(Debug)]
 pub struct ArmCommunicationInterface<S: ArmDebugState> {
-    probe: Box<dyn DapProbe>,
+    probe: Option<Box<dyn DapProbe>>,
```

Comment:
Why does this have to be an Option? I liked the previous version where the DP is unknown better. I feel like an ArmCommunicationInterface should always have a probe.

Reply:
Honestly, I would rather mem::replace a FakeProbe in than have the Option which requires us to do weird access everywhere.
I think the Option is really ugly :/

Reply:
Which was my first suggestion, too :)

Reply:
Yeah I did not get that part, sorry.

Reply:
The issue is that the `Drop` implementation will then be called with the `FakeProbe`, so we would need a check to see if the probe is still there or not. That is a lot cleaner with the Option, otherwise we would have to add an additional field, or check the actual type.

As mentioned in the PR, I'm also not very happy with this...

Reply:
Tried it out here: https://github.com/Tiwalun/probe-rs/commit/f0d2cf33ddc7fbc645c4acaa462dd4178f5cbdc0

Reply:
Do I miss something? `dropped` seems to never be read. What is it good for?

Reply:
Forgot to actually use it, updated the branch: https://github.com/probe-rs/probe-rs/commit/86a4c0c9e9183b809b802131f6bab1d0451e3d38

It has to be used in the `drop` implementation, otherwise we try to drop the fake probe.

Reply:
Is this actually a problem? We just call an empty disconnect function on the `FakeProbe`. I think that's actually fine. Also, maybe we can name it `DroppedProbe` or similar to make it a bit more obvious :)

Reply:
The disconnect function is not part of the probe, it is a sequence. And as soon as that sequence wants to read anything, it will break.

I don't really understand the problem with the `Option` to be honest, it clearly shows what's happening. If we use the `FakeProbe` trick, you just start pretending that the probe is always valid, even when it's not.